Add a dry-run summary mode for TableRebalance which only returns a summary of the dry-run results

somandal opened the pull request:
This PR adds a `summary` option to the TableRebalance API which is meant to be used with `dryRun`. If `summary` is set to `true` a summary of the dry-run is returned rather than the full dry-run. This summary gives some stats about changes that will occur during the rebalance, such as:

- Total number of segments to be moved
- Unique and total number of segments
- Replication factor
- Number of servers
- Server map to segment add/remove/unchanged information
- Average segment size
- Number of servers getting segments added
- Total data size being moved (calculated based on average segment size - based on TableSize)
- An estimate of how long the move can take (based on estimates to download and process each segment) - we should decide on a good throughput estimate as part of this PR review (left a TODO where the code is relevant)

Today the dry-run output can be very large, and can be difficult to make sense of in terms of the changes occurring. It can also be difficult to display the full output. For now `summary` is not appended to `dryRun` without `summary` enabled, but this can be added if it makes sense to add a summary even for the usual result return.

Note: This PR will have conflicts with https://github.com/apache/pinot/pull/15029 and will need to be rebased once that is merged.

Sample JSON summary [NOTE - please see later comments for updated summary based on review comments]:
```
{
  "totalSegmentsToBeMoved" : 6,
  "numServersGettingNewSegments" : 1,
  "estimatedAverageSegmentSizeInBytes" : 1690546,
  "totalEstimatedDataToBeMovedInBytes" : 10143276,
  "totalEstimatedTimeToMoveDataInSecs" : 0.09673381805419921,
  "numServers" : {
    "_existingValue" : 1,
    "_newValue" : 2
  },
  "replicationFactor" : {
    "_existingValue" : 1,
    "_newValue" : 1
  },
  "numUniqueSegments" : {
    "_existingValue" : 12,
    "_newValue" : 12
  },
  "numTotalSegments" : {
    "_existingValue" : 12,
    "_newValue" : 12
  },
  "serverSegmentChangeInfo" : {
    "Server_localhost_22004" : {
      "_totalNewSegments" : 6,
      "_totalExistingSegments" : 0,
      "_segmentsAdded" : 6,
      "_segmentsDeleted" : 0,
      "_segmentsUnchanged" : 0
    },
    "Server_localhost_22001" : {
      "_totalNewSegments" : 6,
      "_totalExistingSegments" : 12,
      "_segmentsAdded" : 0,
      "_segmentsDeleted" : 6,
      "_segmentsUnchanged" : 6
    }
  }
}
```

cc @Jackie-Jiang @klsince @deepthi912 @npawar

npawar commented:
> This PR adds a `summary` option to the TableRebalance API which is meant to be used with `dryRun`. If `summary` is set to `true` a summary of the dry-run is returned rather than the full dry-run. This summary gives some stats about changes that will occur during the rebalance, such as:
> 
> * Total number of segments to be moved
> * Unique and total number of segments
> * Replication factor
> * Number of servers
> * Server map to segment add/remove/unchanged information
> * Average segment size
> * Number of servers getting segments added
> * Total data size being moved (calculated based on average segment size - based on TableSize)
> * An estimate of how long the move can take (based on estimates to download and process each segment) - we should decide on a good throughput estimate as part of this PR review (left a TODO where the code is relevant)
> 
> Today the dry-run output can be very large, and can be difficult to make sense of in terms of the changes occurring. It can also be difficult to display the full output. For now `summary` is not appended to `dryRun` without `summary` enabled, but this can be added if it makes sense to add a summary even for the usual result return.
> 
> Note: This PR will have conflicts with #15029 and will need to be rebased once that is merged.
> 
> Sample JSON summary:
> 
> ```
> {
>   "totalSegmentsToBeMoved" : 6,
>   "numServersGettingNewSegments" : 1,
>   "estimatedAverageSegmentSizeInBytes" : 1690546,
>   "totalEstimatedDataToBeMovedInBytes" : 10143276,
>   "totalEstimatedTimeToMoveDataInSecs" : 0.09673381805419921,
>   "numServers" : {
>     "_existingValue" : 1,
>     "_newValue" : 2
>   },
>   "replicationFactor" : {
>     "_existingValue" : 1,
>     "_newValue" : 1
>   },
>   "numUniqueSegments" : {
>     "_existingValue" : 12,
>     "_newValue" : 12
>   },
>   "numTotalSegments" : {
>     "_existingValue" : 12,
>     "_newValue" : 12
>   },
>   "serverSegmentChangeInfo" : {
>     "Server_localhost_22004" : {
>       "_totalNewSegments" : 6,
>       "_totalExistingSegments" : 0,
>       "_segmentsAdded" : 6,
>       "_segmentsDeleted" : 0,
>       "_segmentsUnchanged" : 0
>     },
>     "Server_localhost_22001" : {
>       "_totalNewSegments" : 6,
>       "_totalExistingSegments" : 12,
>       "_segmentsAdded" : 0,
>       "_segmentsDeleted" : 6,
>       "_segmentsUnchanged" : 6
>     }
>   }
> }
> ```
> 
> cc @Jackie-Jiang @klsince @deepthi912 @npawar

A few comments on the summary. Feel free to just take those that make sense for this first iteration:
1. along with numServers, it might be useful to see the list there of existing and new, so operator can confirm that their tagging / untagging is effective (so a servers added / removed / unchanged ?)
2. how about showing what tenant tag we're operating with? so summarize the tags from tenants, completed, tier, pools
3. what does num unique segments mean? same with numTotalSegments, didn't follow what existing/new value means in context of rebalance. Perhaps having a description field within the sections will help.
4. in the server to stats map for 22001, if it started with 12 and 6 are moving, shouldn't totalNewSegments be 0?
5. this payload will get pretty extensive over time. wondering if we should take time to do some more top level categorization - segments related info, servers related info, generic info

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+      long totalDataPerServerInBytes = totalEstimatedDataToBeMovedInBytes / numServersGettingDataAdded;
+      // TODO: Pick a good threshold to calculate estimated time to rebalance. For now assume 100 MB/s data download
+      //       + process rate
+      estimatedTimeToRebalanceInSec = ((double) totalDataPerServerInBytes) / (100.0D * 1024.0D * 1024.0D);
```
I think maybe we skip this field, as it's not easy to estimate this accurately because the cluster setup can vary a lot. The summary contains the total amount of segments and data bytes a server may get now, so the user can estimate the duration for their envs accordingly.

somandal replied:
yeah, this is a tricky field for sure. I've removed it for now, but added a TODO for adding it back once we have a better understanding of a good estimate. This field was an ask for this summary PR.

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -291,43 +305,58 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+    RebalanceSummaryResult summaryResult = null;
+    if (summary) {
+      summaryResult = calculateDryRunSummary(currentAssignment, targetAssignment, tableNameWithType, rebalanceJobId);
+    }
```
how about calculating the summary after this whole if-block, so we save the cost from calculating it when the table is already balanced?
```
if (segmentAssignmentUnchanged) {
..
}
```

somandal replied:
I think it might still be useful to see the summary in case the table being balanced is not expected? wdyt?
I haven't moved it yet, but let me know if you still think it's not worth calculating the summary for this scenario and I can move it down and adjust the tests.

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+    } catch (InvalidConfigException e) {
+      String errMsg = String.format("Caught exception while trying to fetch table size details for table: %s",
+          tableNameWithType);
+      LOGGER.error(errMsg, e);
```
inline `LOGGER.error("Caught.... {}", tableWithType, e)`? as errMsg is not used anywhere else.

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+      existingReplicationFactor = entrySet.getValue().size();
+      for (Map.Entry<String, String> segmentEntrySet : entrySet.getValue().entrySet()) {
+        existingServersToSegmentMap.putIfAbsent(segmentEntrySet.getKey(), new HashSet<>());
+        existingServersToSegmentMap.get(segmentEntrySet.getKey()).add(entrySet.getKey());
```
nit: existingServersToSegmentMap.computeIfAbsent().add()

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+      newReplicationFactor = entrySet.getValue().size();
+      for (Map.Entry<String, String> segmentEntrySet : entrySet.getValue().entrySet()) {
+        newServersToSegmentMap.putIfAbsent(segmentEntrySet.getKey(), new HashSet<>());
+        newServersToSegmentMap.get(segmentEntrySet.getKey()).add(entrySet.getKey());
```
computeIfAbsent().add()? and just loop over keys (server names)?

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+
+    for (Map.Entry<String, Map<String, String>> entrySet : currentAssignment.entrySet()) {
+      existingReplicationFactor = entrySet.getValue().size();
+      for (Map.Entry<String, String> segmentEntrySet : entrySet.getValue().entrySet()) {
```
looks like we just need the keys (i.e. server names) in this loop

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+        = new RebalanceSummaryResult.RebalanceChangeInfo(existingReplicationFactor, newReplicationFactor);
+
+    int existingNumServers = existingServersToSegmentMap.keySet().size();
+    int newNumServers = newServersToSegmentMap.keySet().size();
```
nit: simply map.size()

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+    int numServersGettingDataAdded = 0;
+    for (Map.Entry<String, Set<String>> entry : newServersToSegmentMap.entrySet()) {
+      String server = entry.getKey();
+      Set<String> segmentMap = entry.getValue();
```
nit: segments or segmentSet

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+      int totalExistingSegments = 0;
+      RebalanceSummaryResult.ServerStatus serverStatus = RebalanceSummaryResult.ServerStatus.ADDED;
+      if (existingServersToSegmentMap.containsKey(server)) {
+        totalExistingSegments = existingServersToSegmentMap.get(server).size();
```
nit: use a var to keep existingServersToSegmentMap.get(server), as this map.get() is called 3 times in the loop. Or get value of map.get() and check if it's null for the if-check

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+      int totalNewSegments = segmentMap.size();
+
+      Set<String> newSegmentList = new HashSet<>(segmentMap);
+      Set<String> existingSegmentList = new HashSet<>();
```
nit: perhaps remove `List` suffix from the var name if using Set<> for bit less confusion

somandal replied:
done

klsince commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+    }
+    try {
+      TableSizeReader.TableSubTypeSizeDetails tableSizeDetails =
+          _tableSizeReader.getTableSubtypeSize(tableNameWithType, 30_000);
```
perhaps add LOGGER.info() before/after this method, to know how much time controller may spend getting the table size, as I assume it'd take a while for large tables

and maybe leave a TODO to make 30s configurable later

somandal replied:
done

npawar commented:
> Updated summary:
> 
> Adding a server:
> 
> ```
> {
>   "serverInfo" : {
>     "numServersGettingNewSegments" : 1,
>     "numServers" : {
>       "valueBeforeRebalance" : 1,
>       "expectedValueAfterRebalance" : 2
>     },
>     "serverSegmentChangeInfo" : {
>       "Server_localhost_22004" : {
>         "serverStatus" : "ADDED",
>         "totalSegmentsAfterRebalance" : 6,
>         "totalSegmentsBeforeRebalance" : 0,
>         "segmentsAdded" : 6,
>         "segmentsDeleted" : 0,
>         "segmentsUnchanged" : 0,
>         "tagList" : [ "DefaultTenant_OFFLINE", "DefaultTenant_REALTIME" ]
>       },
>       "Server_localhost_22001" : {
>         "serverStatus" : "UNCHANGED",
>         "totalSegmentsAfterRebalance" : 6,
>         "totalSegmentsBeforeRebalance" : 12,
>         "segmentsAdded" : 0,
>         "segmentsDeleted" : 6,
>         "segmentsUnchanged" : 6,
>         "tagList" : [ "DefaultTenant_OFFLINE", "DefaultTenant_REALTIME" ]
>       }
>     }
>   },
>   "segmentInfo" : {
>     "totalSegmentsToBeMoved" : 6,
>     "estimatedAverageSegmentSizeInBytes" : 1690546,
>     "totalEstimatedDataToBeMovedInBytes" : 10143276,
>     "totalEstimatedTimeToMoveDataInSecs" : 0.09673381805419921,
>     "replicationFactor" : {
>       "valueBeforeRebalance" : 1,
>       "expectedValueAfterRebalance" : 1
>     },
>     "numSegmentsInSingleReplica" : {
>       "valueBeforeRebalance" : 12,
>       "expectedValueAfterRebalance" : 12
>     },
>     "numSegmentsAcrossAllReplicas" : {
>       "valueBeforeRebalance" : 12,
>       "expectedValueAfterRebalance" : 12
>     }
>   }
> }
> ```
> 
> Removing a server:
> 
> ```
> {
>   "serverInfo" : {
>     "numServersGettingNewSegments" : 1,
>     "numServers" : {
>       "valueBeforeRebalance" : 2,
>       "expectedValueAfterRebalance" : 1
>     },
>     "serverSegmentChangeInfo" : {
>       "Server_localhost_22004" : {
>         "serverStatus" : "REMOVED",
>         "totalSegmentsAfterRebalance" : 0,
>         "totalSegmentsBeforeRebalance" : 6,
>         "segmentsAdded" : 0,
>         "segmentsDeleted" : 6,
>         "segmentsUnchanged" : 0,
>         "tagList" : [ ]
>       },
>       "Server_localhost_22001" : {
>         "serverStatus" : "UNCHANGED",
>         "totalSegmentsAfterRebalance" : 12,
>         "totalSegmentsBeforeRebalance" : 6,
>         "segmentsAdded" : 6,
>         "segmentsDeleted" : 0,
>         "segmentsUnchanged" : 6,
>         "tagList" : [ "DefaultTenant_OFFLINE", "DefaultTenant_REALTIME" ]
>       }
>     }
>   },
>   "segmentInfo" : {
>     "totalSegmentsToBeMoved" : 6,
>     "estimatedAverageSegmentSizeInBytes" : 1690546,
>     "totalEstimatedDataToBeMovedInBytes" : 10143276,
>     "totalEstimatedTimeToMoveDataInSecs" : 0.09673381805419921,
>     "replicationFactor" : {
>       "valueBeforeRebalance" : 1,
>       "expectedValueAfterRebalance" : 1
>     },
>     "numSegmentsInSingleReplica" : {
>       "valueBeforeRebalance" : 12,
>       "expectedValueAfterRebalance" : 12
>     },
>     "numSegmentsAcrossAllReplicas" : {
>       "valueBeforeRebalance" : 12,
>       "expectedValueAfterRebalance" : 12
>     }
>   }
> }
> ```
> 
> > A few comments on the summary. Feel free to just take those that make sense for this first iteration:
> > 
> > 1. along with numServers, it might be useful to see the list there of existing and new, so operator can confirm that their tagging / untagging is effective (so a servers added / removed / unchanged ?)
> > 2. how about showing what tenant tag we're operating with? so summarize the tags from tenants, completed, tier, pools
> > 3. what does num unique segments mean? same with numTotalSegments, didn't follow what existing/new value means in context of rebalance. Perhaps having a description field within the sections will help.
> > 4. in the server to stats map for 22001, if it started with 12 and 6 are moving, shouldn't totalNewSegments be 0?
> > 5. this payload will get pretty extensive over time. wondering if we should take time to do some more top level categorization - segments related info, servers related info, generic info
> 
> @npawar responses regarding the above are below:
> 
> 1. done, added a server status for each server
> 2. I've added the tenant tags for each server - is this what you meant? Or do you think just dumping the list of all tenant tags across servers is good enough? Or did you want finer categorization based on whether it's OFFLINE/CONSUMING/COMPLETED/TIER etc? For more categorization, we can add it as part of a future change?
> 3. by unique segments -> i meant number of segments per replica. I've updated the name of the field in the new summary. Total segments was the total number of segments across all replicas. Existing was the value before rebalance, and the value after expected rebalance. I've updated the fields in the summary. Let me know if that makes more sense or if we still need a description field
> 4. here `totalNewSegments` was total segments hosted after the rebalance completes. So 22001 was losing 6 of the existing 12 segments. I've updated the field to better explain this.
> 5. I've attempted a categorization into server vs. segments. Please take a look and let me know what you think

points 3,4,5 look great.
point 2 - sure we can take this up later
For 1, the tag in each server's section is helpful. What I was suggesting though, is that we provide the list right on top (Essentially 4 lists called serversGettingNewSegments, serverAdded, serverRemoved, serverUnchanged) as when we'll get to clusters with 50-100 servers, parsing through 50-100 individual json blurbs per server might get tricky.

somandal commented:
> points 3,4,5 look great.
> point 2 - sure we can take this up later
> For 1, the tag in each server's section is helpful. What I was suggesting though, is that we provide the list right on top (Essentially 4 lists called serversGettingNewSegments, serverAdded, serverRemoved, serverUnchanged) as when we'll get to clusters with 50-100 servers, parsing through 50-100 individual json blurbs per server might get tricky.

Got it, updated the PR to add your suggestion for 1.

Sample summary:

Add server:
```
{
  "serverInfo" : {
    "numServersGettingNewSegments" : 1,
    "numServers" : {
      "valueBeforeRebalance" : 1,
      "expectedValueAfterRebalance" : 2
    },
    "serversAdded" : [ "Server_localhost_22004" ],
    "serversRemoved" : [ ],
    "serversUnchanged" : [ "Server_localhost_22001" ],
    "serversGettingNewSegments" : [ "Server_localhost_22004" ],
    "serverSegmentChangeInfo" : {
      "Server_localhost_22004" : {
        "serverStatus" : "ADDED",
        "totalSegmentsAfterRebalance" : 6,
        "totalSegmentsBeforeRebalance" : 0,
        "segmentsAdded" : 6,
        "segmentsDeleted" : 0,
        "segmentsUnchanged" : 0,
        "tagList" : [ "DefaultTenant_OFFLINE", "DefaultTenant_REALTIME" ]
      },
      "Server_localhost_22001" : {
        "serverStatus" : "UNCHANGED",
        "totalSegmentsAfterRebalance" : 6,
        "totalSegmentsBeforeRebalance" : 12,
        "segmentsAdded" : 0,
        "segmentsDeleted" : 6,
        "segmentsUnchanged" : 6,
        "tagList" : [ "DefaultTenant_OFFLINE", "DefaultTenant_REALTIME" ]
      }
    }
  },
  "segmentInfo" : {
    "totalSegmentsToBeMoved" : 6,
    "maxSegmentsAddedToASingleServer" : 6,
    "estimatedAverageSegmentSizeInBytes" : 1690546,
    "totalEstimatedDataToBeMovedInBytes" : 10143276,
    "replicationFactor" : {
      "valueBeforeRebalance" : 1,
      "expectedValueAfterRebalance" : 1
    },
    "numSegmentsInSingleReplica" : {
      "valueBeforeRebalance" : 12,
      "expectedValueAfterRebalance" : 12
    },
    "numSegmentsAcrossAllReplicas" : {
      "valueBeforeRebalance" : 12,
      "expectedValueAfterRebalance" : 12
    }
  }
}
```

Remove server:
```
{
  "serverInfo" : {
    "numServersGettingNewSegments" : 1,
    "numServers" : {
      "valueBeforeRebalance" : 2,
      "expectedValueAfterRebalance" : 1
    },
    "serversAdded" : [ ],
    "serversRemoved" : [ "Server_localhost_22004" ],
    "serversUnchanged" : [ "Server_localhost_22001" ],
    "serversGettingNewSegments" : [ "Server_localhost_22001" ],
    "serverSegmentChangeInfo" : {
      "Server_localhost_22004" : {
        "serverStatus" : "REMOVED",
        "totalSegmentsAfterRebalance" : 0,
        "totalSegmentsBeforeRebalance" : 6,
        "segmentsAdded" : 0,
        "segmentsDeleted" : 6,
        "segmentsUnchanged" : 0,
        "tagList" : [ ]
      },
      "Server_localhost_22001" : {
        "serverStatus" : "UNCHANGED",
        "totalSegmentsAfterRebalance" : 12,
        "totalSegmentsBeforeRebalance" : 6,
        "segmentsAdded" : 6,
        "segmentsDeleted" : 0,
        "segmentsUnchanged" : 6,
        "tagList" : [ "DefaultTenant_OFFLINE", "DefaultTenant_REALTIME" ]
      }
    }
  },
  "segmentInfo" : {
    "totalSegmentsToBeMoved" : 6,
    "maxSegmentsAddedToASingleServer" : 6,
    "estimatedAverageSegmentSizeInBytes" : 1690546,
    "totalEstimatedDataToBeMovedInBytes" : 10143276,
    "replicationFactor" : {
      "valueBeforeRebalance" : 1,
      "expectedValueAfterRebalance" : 1
    },
    "numSegmentsInSingleReplica" : {
      "valueBeforeRebalance" : 12,
      "expectedValueAfterRebalance" : 12
    },
    "numSegmentsAcrossAllReplicas" : {
      "valueBeforeRebalance" : 12,
      "expectedValueAfterRebalance" : 12
    }
  }
}
```

Also note that I've removed the time estimates for now as recommended by @klsince as it is hard to come up with a good estimate (different environments may have different speeds). We can perhaps come up with some heuristic based on some real rebalances if that makes sense in the future.

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/RebalanceSummaryResult.java:
```diff
@@ -0,0 +1,325 @@
+      return _numSegmentsAcrossAllReplicas;
+    }
+  }
+
```
good point, moved it

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,146 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+      Set<String> segmentSet = entry.getValue();
+      int totalNewSegments = segmentSet.size();
+
+      Set<String> newSegmentList = new HashSet<>(segmentSet);
```
oops, done

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -559,6 +588,146 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+    } catch (InvalidConfigException e) {
+      LOGGER.error("Caught exception while trying to fetch table size details for table: {}", tableNameWithType, e);
+    }
+    LOGGER.info("Fetched the table size for rebalance summary for table: {}", tableNameWithType);
```
done
